src/Libraries/Microsoft.Extensions.TimeProvider.Testing/FakeTimeProvider.cs

```diff
@@ -240,15 +243,29 @@ private void WakeWaiters()
                 return;
             }
 
+            var oldTicks = _now.Ticks;
+
             // invoke the callback
             candidate.InvokeCallback();
 
+            var newTicks = _now.Ticks;
+
             // see if we need to reschedule the waiter
             if (candidate.Period > 0)
             {
                 // update the waiter's state
-                candidate.ScheduledOn = _now.Ticks;
-                candidate.WakeupTime += candidate.Period;
+                candidate.ScheduledOn = newTicks;
+
+                if (oldTicks != newTicks)
+                {
+                    // time changed while in the callback, readjust the wake time accordingly
+                    candidate.WakeupTime = newTicks + candidate.Period;
+                }
+                else
+                {
+                    // move on to the next period
+                    candidate.WakeupTime += candidate.Period;
+                }
```

Comment:
Good solution. This is safe because the current thread that is invoking the waiters/callback is the only one that can change _ now, since it is protected by the lock, and it is only that thread that can reenter the lock. So a change to _now could only come from the same thread.

Is my understanding correct?

Reply:
Exactly :-)